multiplexer/abci/multiplexer.go

```diff
@@ -0,0 +1,662 @@
+	// chainID is required as it needs to be propagated to the ABCI V1 connection.
+	chainID string
+	// cmNode is the comet node which has been created. A reference is required in order to establish
+	// a local connection to it.
+	cmNode *node.Node
+	// versions is a list of versions which contain all embedded binaries.
+	versions Versions
+	// conn is a grpc client connection and used when creating remote ABCI connections.
+	conn *grpc.ClientConn
+	// cleanupFns is a list of functions which should execute upon cleanup of the multiplexer.
+	// any returned errors are logged.
+	cleanupFns []func() error
+	// ctx is the context which is passed to the comet, grpc and api server starting functions.
+	ctx context.Context
+	// g is the waitgroup to which the comet, grpc and api server init functions are added to.
+	g *errgroup.Group
+}
+
+// NewVersions returns a list of versions sorted by app version.
+func NewVersions(v ...Version) (Versions, error) {
+	versions := Versions(v)
+	if err := versions.Validate(); err != nil {
+		return nil, err
+	}
+	return versions.Sorted(), nil
+}
+
+// NewMultiplexer creates a new Multiplexer.
+func NewMultiplexer(svrCtx *server.Context, svrCfg serverconfig.Config, clientCtx client.Context, appCreator servertypes.AppCreator, versions Versions, chainID string, applicationVersion uint64) (*Multiplexer, error) {
+	if err := versions.Validate(); err != nil {
+		return nil, fmt.Errorf("invalid versions: %w", err)
+	}
+
+	mp := &Multiplexer{
+		svrCtx:        svrCtx,
+		svrCfg:        svrCfg,
+		clientContext: clientCtx,
+		appCreator:    appCreator,
+		logger:        svrCtx.Logger.With("multiplexer"),
+		nativeApp:     nil, // app will be initialized if required by the multiplexer.
+		versions:      versions,
+		chainID:       chainID,
+		appVersion:    applicationVersion,
+		cleanupFns:    make([]func() error, 0),
+	}
+
+	return mp, nil
+}
+
+// isRunningNativeApp checks if the native application is currently running.
+func (m *Multiplexer) isNativeApp() bool {
+	return m.nativeApp != nil
+}
+
+// isEmbeddedApp checks if an embedded application is currently running.
+func (m *Multiplexer) isEmbeddedApp() bool {
+	return !m.isNativeApp()
+}
+
+// isGrpcOnly checks if the GRPC-only mode is enabled using the configuration flag.
+func (m *Multiplexer) isGrpcOnly() bool {
+	return m.svrCtx.Viper.GetBool(flagGRPCOnly)
+}
+
+// registerCleanupFn enables the registration of additional cleanup functions that get called during Cleanup
+func (m *Multiplexer) registerCleanupFn(cleanUpFn func() error) {
+	m.cleanupFns = append(m.cleanupFns, cleanUpFn)
+}
+
+func (m *Multiplexer) Start() error {
+	m.g, m.ctx = getCtx(m.svrCtx, true)
+
+	emitServerInfoMetrics()
+
+	// startApp starts the underlying application, either native or embedded.
+	if err := m.startApp(); err != nil {
+		return err
+	}
+
+	if m.isGrpcOnly() {
+		m.logger.Info("starting node in gRPC only mode; CometBFT is disabled")
+		m.svrCfg.GRPC.Enable = true
+	} else {
```

Comment:
this case is always confusing to me...maybe we should add a comment or something. this is for starting an app in standalone mode, but with no comet underneath? is this ever something we want or need?

Reply:
yes, it is useful when willing to just query state, you download the db, start in grpc only and you are done.